### PR TITLE
feat: Add new private_key & passphrase based authentication method

### PR DESCRIPTION
**Pull request by haleemur**
# Description

The tap can now connect to snowflake using a provided private key and passphrase. The private key can be provided as a string or the path to the private key file can be provided.

Fixes an edge-case handling where schemas needing quotes (example: `"MySpecial.SchemaName"`) are properly handled during discovery.

Speeds up discovery slightly when `tables` is provided. The speed up is achieved by only inspecting relevant schemas which can be derived by parsing each table within `tables`.

Introduces 3 new config parameters to support key based authentication and since there are 2 ways to authenticate, `password` is now an optional parameter. The new parameters are:
- `private_key`
- `private_key_path`
- `private_key_passphrase`

resolves: https://github.com/MeltanoLabs/tap-snowflake/issues/19

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -29,7 +29,13 @@ environment variable is set either in the terminal context or in the `.env` file
 
 ### Source Authentication and Authorization
 
-Standard `username` and `password` auth is supported.
+This tap supports two authentication methods:
+- Standard `username` and `password` auth
+- [Key Pair Authentication](https://docs.snowflake.com/en/user-guide/key-pair-auth)
+
+### Key Pair Authentication
```
Comment (edgarrmondragon) [inline README.md:36]:
```suggestion
#### Key Pair Authentication
```

File: meltano.yml [hunk 2]
```diff
@@ -14,6 +14,11 @@ plugins:
     - name: user
     - name: password
       kind: password
+    - name: private_key
+      kind: password
```
Comment (edgarrmondragon) [inline meltano.yml:18]:
```suggestion
      kind: password
      sensitive: true
```

File: meltano.yml [hunk 3]
```diff
@@ -14,6 +14,11 @@ plugins:
...
+    - name: private_key_path
+    - name: private_key_passphrase
+      kind: password
```
Comment (edgarrmondragon) [inline meltano.yml:21]:
```suggestion
      kind: password
      sensitive: true
```

File: tap_snowflake/client.py [hunk 4]
```diff
@@ -74,15 +76,59 @@ class TableProfile:
...
+    @property
+    def auth_method(self):
+        """
+        Validate & return the authentication method based on config.
+
+        Cache computed auth_method to attribute `_auth_method`.
+        """
+        try:
+            return self._auth_method
+        except AttributeError:
```
Comment (edgarrmondragon) [inline tap_snowflake/client.py:112]: Would there be any downsides to instead do something like

```python
class SnowflakeConnector(SQLConnector):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._auth_method: str | None = None

    def auth_method(self) -> str:
        if self._auth_method:
            return self._auth_method
        ...
```

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]

File: tap_snowflake/client.py [hunk 5]
```diff
@@ -12,11 +12,13 @@
 from typing import Any, Iterable, List, Tuple
 from uuid import uuid4
 import datetime
-
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import serialization
```
[the remark this answers is not in the document]
Comment (haleemur-infostrux) [reply tap_snowflake/client.py:16]: good suggestion. when this code worked, i just assumed cryptography was built in. definitely don't want to introduce a transitive dependency.

File: tap_snowflake/client.py (hunk 4 above)
Comment (haleemur-infostrux) [reply tap_snowflake/client.py:112]: No downsides as far as I can see, and I like yours as it avoids using exceptions for control flow, but it requires over-riding the init method. How do you feel about the following:

```
class SnowflakeConnector(SqlConnector):
  _auth_method: str | None = None

  def auth_method(self) -> str:
     if not self._auth_method:
       ...
     return self._auth_method
```

Comment (edgarrmondragon) [reply tap_snowflake/client.py]: > but it requires over-riding the init method

For my own curiosity, are there any downsides to doing that?

Comment (haleemur) [reply tap_snowflake/client.py]: no, as far as i can tell, the only downsides are additional lines one has to type, and every subsequent reader has to read - no downsides to performance or code organization.

actually, we can make the implementation even cleaner. New commit coming in a few minutes.

Comment (haleemur) [reply tap_snowflake/client.py]: @edgarrmondragon the new implementation uses `functools.cached_property`, thus we don't require declaring either a class or an instance variable to cache the value of this method evaluation. Also the if condition in the method can be removed making that code tighter as well.

**Review by edgarrmondragon — APPROVED** [review]
LGTM!

**Comment by edgarrmondragon** [comment]
Thanks @haleemur!

**Pull request merged**